feat(docs): add copy code snippet button to code blocks

joannalauu opened the pull request:
## PR Checklist

- [x] Commit message follows our [contributing guidelines](https://github.com/graycoreio/daffodil/blob/develop/CONTRIBUTING.md#commit)
- [x] Tests added/updated (for bug fixes/features)
- [ ] Documentation added/updated (for bug fixes/features)

## PR Type
<!-- Select the relevant type(s) -->

- [ ] Bug fix
- [x] Feature
- [ ] Style update
- [ ] Refactor
- [ ] Test
- [ ] Build
- [ ] CI
- [ ] Docs
- [ ] Performance
- [ ] Other (please describe)

## Current behavior
<!-- Describe the current behavior and link to relevant issue -->

Fixes: #3949 

## New behavior
Adds a copy button to the code blocks in the daffio guides. Code block is copied to clipboard on click.

## Breaking change?

- [ ] Yes
- [x] No

<!-- If yes, describe impact and migration path -->

## Additional context

https://github.com/user-attachments/assets/efdf593d-4783-4008-855b-7c19c1f3353e

joannalauu committed:
works but only appear on page refresh

joannalauu committed:
copy button works!!

joannalauu committed:
improve code safety

joannalauu committed:
fix tests

joannalauu committed:
remove unnecessary imports

joannalauu commented:
Please let me know if I should change the code structure in the docs package – I wasn't sure how to organize the files because there wasn't an existing structure to base off of. Also let me know if I should move the addCopyButtonsToCodeBlocks and cleanupButtons functions outside of guides-content.component.ts and then just import them in.

griest024 reviewed (dismissed):
So ideally we want to add the HTML content in dgeni as I described here: https://github.com/graycoreio/daffodil/issues/3949#issuecomment-3308764140. The way you've done it in daffio here is not [hydration safe](https://angular.dev/guide/hydration#direct-dom-manipulation)

The tricky part of that is that you won't be able to use Angular components. So unless we want to use custom elements at this time (@damienwebdev we don't, right?) you'll need to convert `DaffDocsCopyButtonComponent` to pure HTML/JS/CSS and add it to the content in dgeni.

EDIT: @sunray4 actually standby, it might be hydration safe because of `afterRenderEffect`

griest024 commented:
@sunray4 thanks for your contribution! I hope your exams went well!

griest024 commented on apps/daffio/src/app/docs/guides/components/guides-content/guides-content.component.ts:
```diff
@@ -29,12 +36,74 @@ export class DaffioDocsGuidesContentComponent implements DaffioDocsDynamicConten
+    afterRenderEffect(() => {
+      this.doc();
+      this.addCopyButtonsToCodeBlocks();
+    });
```
Prefer explicitly specifying the phase here. Can you get it working without using `mixedReadWrite`?

> You should prefer using the read and write phases over the `earlyRead` and `mixedReadWrite` phases when possible, to avoid performance degradation.

https://angular.dev/api/core/afterRenderEffect

griest024 commented on apps/daffio/src/app/docs/guides/components/guides-content/guides-content.component.ts:
```diff
@@ -29,12 +36,74 @@ export class DaffioDocsGuidesContentComponent implements DaffioDocsDynamicConten
+      // Add copy button to the code block
+      pre.appendChild(buttonElement);
+
+      this.buttonRefs.push(buttonRef);
```
We'll need to call this in every doc kind component so this should be extracted out so that it is reusable.

joannalauu replied:
Sounds good, in this case I'll just put the addCopyButtonsToCodeBlocks and cleanup functionality into a service inside the doc package – may I confirm if this works?

griest024 replied:
yes but be careful about the injection hierarchy of the service. It must inherit the injector of the component to get the correct `elementRef`, etc.

joannalauu committed:
specify write phase for afterRenderEffect

joannalauu committed:
moved append button to code block functionality into service

joannalauu committed:
add spec tests for service

joannalauu committed:
remove placeholder spec tests

griest024 commented on apps/daffio/src/app/docs/guides/components/guides-content/guides-content.component.ts:
```diff
@@ -29,12 +36,30 @@ export class DaffioDocsGuidesContentComponent implements DaffioDocsDynamicConten
+    });
+
+    afterRenderEffect({
+      write: () => {
```
This callback should accept a cleanup arg. I forget the exact API but you shouldn't need to touch `DestroyRef` directly

joannalauu committed:
use cleanup callback instead of destroyref

griest024 commented on apps/daffio/src/app/docs/guides/components/guides-content/guides-content.component.ts:
```diff
@@ -29,12 +35,29 @@ export class DaffioDocsGuidesContentComponent implements DaffioDocsDynamicConten
+    afterRenderEffect({
+      write: (onCleanup) => {
+        this.doc();
+        this.copyButtonService.addCopyButtonsToCodeBlocks(this.elementRef.nativeElement, this.viewContainerRef);
```
You can directly inject the args here, the constructor is a valid injection context.

griest024 reviewed (dismissed):
Nice code! :clap: you need to add this to all the other doc kinds now: API, design, packages, etc. Take care that the copy button is not added to the source code block (which all the API pages have)

xelaint commented on libs/docs/src/copy-button/service/copy-code-block.service.ts:
```diff
@@ -0,0 +1,61 @@
+
+      const textContent = code.textContent || '';
+
+      pre.style.position = 'relative';
```
We should just update the `pre` style in `DaffArticleComponent` to include this style.

xelaint commented on libs/docs/src/copy-button/service/copy-code-block.service.ts:
```diff
@@ -0,0 +1,61 @@
+      const buttonElement = buttonRef.location.nativeElement;
+
+      const innerButton = buttonElement.querySelector('button');
+      if (innerButton) {
```
I don't think adding a class to innerButton on copy is necessary. The positioning can just be applied to the host component.

xelaint commented on libs/docs/src/copy-button/copy-button/copy-button.component.scss:
```diff
@@ -0,0 +1,20 @@
+.daff-docs-copy-button {
```
This style file should just be:

```
:host {
	display: block;
	position: absolute;
	top: 0.5rem;
	right: 0.5rem;
	z-index: 1;
}
```

state styles (hover and focus) should be handled by the button component.

xelaint commented on libs/docs/src/copy-button/copy-button/copy-button.component.html:
```diff
@@ -0,0 +1,13 @@
+  daff-icon-button
+  size="sm"
+  type="button"
+  class="daff-docs-copy-button"
```
class can be removed

xelaint commented on libs/docs/src/copy-button/copy-button/copy-button.component.html:
```diff
@@ -0,0 +1,13 @@
+  size="sm"
+  type="button"
+  class="daff-docs-copy-button"
+  [attr.aria-label]="copied() ? 'Copied to clipboard' : 'Copy to clipboard'"
```
move logic to component

xelaint commented on libs/docs/src/copy-button/service/copy-code-block.service.ts:
```diff
@@ -0,0 +1,61 @@
+import { DaffDocsCopyButtonComponent } from '../copy-button/copy-button.component';
+
+@Injectable()
+export class CodeBlockCopyButtonService {
```
Prefix this service with `DaffDocs`

xelaint commented on libs/docs/src/copy-button/copy-button/copy-button.component.ts:
```diff
@@ -0,0 +1,66 @@
+  templateUrl: './copy-button.component.html',
+  styleUrls: ['./copy-button.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  standalone: true,
```
can remove `standalone: true;` since it's the default

joannalauu committed:
implement requested changes

joannalauu committed:
add copy button to design, packages, api

joannalauu committed:
update tests

joannalauu committed:
improve styles

joannalauu committed:
remove opacity change

joannalauu committed:
update tests

xelaint commented on libs/docs/src/copy-button/copy-button/copy-button.component.ts:
```diff
@@ -0,0 +1,68 @@
+  styleUrls: ['./copy-button.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  host: {
+    '[attr.aria-label]': 'copied() ? "Copied to clipboard" : "Copy to clipboard"',
```
My requested change was to move the `copied() ? "Copied to clipboard" : "Copy to clipboard"` logic into the component as a function (e.g. DesignLandSidebarViewportComponent), but keep the `aria-label` attribute on the button itself

xelaint reviewed (changes requested):
1. A change you made moved the `<daff-docs-copy-button>` from within `<pre>` to outside of `<pre>`. The button should remain inside of `<pre>`, so you don't need to create an extra div to set a position relative style on it.
2. Design API tabs don't have the copy button

joannalauu commented:
> 1. A change you made moved the `<daff-docs-copy-button>` from within `<pre>` to outside of `<pre>`. The button should remain inside of `<pre>`, so you don't need to create an extra div to set a position relative style on it.

While testing the button, I realized that when the pre element has horizontal scroll, the button moves with scroll. My workaround was to wrap the pre and then put the button in the wrapper – please let me know if there's a better way to do this!!

joannalauu committed:
make afterRenderEffect reactive to tab changes and fix aria-label logic

joannalauu committed:
update aria-label tests

joannalauu commented:
Tests in daffio package component-content.component.spec.ts are failing because of a missing property that is required in DaffPackageGuideDoc. When I checked out to the most recent commit in the upstream develop branch, the test still fails. May I ask how do I resolve this?

xelaint commented:
> > 1. A change you made moved the `<daff-docs-copy-button>` from within `<pre>` to outside of `<pre>`. The button should remain inside of `<pre>`, so you don't need to create an extra div to set a position relative style on it.
> 
> While testing the button, I realized that when the pre element has horizontal scroll, the button moves with scroll. My workaround was to wrap the pre and then put the button in the wrapper – please let me know if there's a better way to do this!!

Ah, good catch. What you did with the div is fine then, no further changes needed.

griest024 commented:
> Tests in daffio package component-content.component.spec.ts are failing because of a missing property that is required in DaffPackageGuideDoc. When I checked out to the most recent commit in the upstream develop branch, the test still fails. May I ask how do I resolve this? 

It's my fault, rebase on develop and you should be good.

damienwebdev commented on libs/docs/package.json:
```diff
@@ -29,10 +29,14 @@
     "@angular/core": "0.0.0-PLACEHOLDER",
     "@daffodil/core": "0.0.0-PLACEHOLDER",
-    "@daffodil/docs-utils": "0.0.0-PLACEHOLDER"
+    "@daffodil/design": "0.0.0-PLACEHOLDER",
```
I was just looking through this PR as I had a moment, I'm just wondering if we couldn't move the `copy` functionality fully into `DaffArticleComponent`. @griest024 @xelaint @sunray4

joannalauu replied:
@damienwebdev Since all code blocks in the daffio should have a copy button, it would make sense for the `copy` functionality to be part of DaffArticleComponent. If there might be use cases of the pre element in DaffArticleComponent where we don't want the element to have the copy button, I can implement a check for a `nocopy` attribute in the pre element which would prevent the copy button from being added. Let me know if that works!

xelaint replied:
I think the option to have a no copy button is good.

joannalauu replied:
Sounds good, I'll start working on implementing this!!

joannalauu committed:
cleanup button wrapper

joannalauu committed:
add check for nocopy attribute

joannalauu committed:
remove add button service from content pages

joannalauu committed:
update description of copy service

joannalauu committed:
working but 2 issues

joannalauu committed:
works but button page is missing

joannalauu committed:
move copy button into article component

joannalauu committed:
update article readme

joannalauu commented:
Hey, I've just finished moving the `copy` functionality into the article component. However, there was one error I couldn't resolve: Once I moved the addCopyButton service into `DaffArticleComponent`, the button page on daffio returns 404. I have attached a clip showing the error. This error didn't exist in my previous implementation, where the addCopyButton service was run in each section of daffio respectively. Please let me know how I can resolve this!

https://github.com/user-attachments/assets/8753243c-87d4-4b55-b1a8-cd053ee258fe

joannalauu committed:
add example

joannalauu commented:
> Hey, I've just finished moving the `copy` functionality into the article component. However, there was one error I couldn't resolve: Once I moved the addCopyButton service into `DaffArticleComponent`, the button page on daffio returns 404. I have attached a clip showing the error. This error didn't exist in my previous implementation, where the addCopyButton service was run in each section of daffio respectively. Please let me know how I can resolve this!

UPDATE: I fixed the error by implementing the button styles from scratch instead of using DaffIconButtonComponent

joannalauu committed:
fix missing button page - implement copy button theming

joannalauu committed:
remove icon button import and update theming

joannalauu commented:
@damienwebdev I've moved the `copy` functionality into `DaffArticleComponent`. May I ask if this implementation is what you were looking for?

Some key changes:
- All code blocks in `DaffArticleComponent` now has a copy button by default. This copy button can be removed by adding a `nocopy` attribute into the `pre` element.
- `DaffDocsCopyButtonComponent` is moved into the `libs/design/article/src` folder and renamed to `DaffArticleCopyButtonComponent`
- `MutationObserver` is used in `DaffArticleComponent` to check for new `pre` elements that may be created within an existing `daff-article`. This may cause unnecessary overhead if a user implements `daff-article` but doesn't include code blocks
- `DaffArticleCopyButtonComponent` is exported as part of `DAFF_ARTICLE_COMPONENTS` from `@daffodil/design/article`
- `DaffArticleCopyButtonComponent` theming is implemented from scratch because importing `DaffIconComponent` into `DaffArticleCopyButtonComponent` breaks www.daff.io/docs/design/components/button


https://github.com/user-attachments/assets/97dce9f6-3f70-49c5-9bbc-416d897ca196

joannalauu committed:
migrate copy button theming into article themes

joannalauu committed:
prevent button from covering code

joannalauu commented:
@griest024 Hey I've made all requested changes, may I ask if there is anything else you'd like me to fix/change?

griest024 commented:
> @griest024 Hey I've made all requested changes, may I ask if there is anything else you'd like me to fix/change?

sorry my review is taking so long. Nag me if I haven't reviewed this by day after tomorrow haha

damienwebdev commented on libs/docs/src/remove-me.spec.ts:
```diff
@@ -1,3 +0,0 @@
-describe('@daffodil/docs | remove me', () => {
-  it('should be removed', () => {});
```
Remove this diff from this PR.

damienwebdev commented on libs/design/article/src/public_api.ts:
```diff
@@ -2,4 +2,5 @@ export { DaffArticleModule } from './article.module';
 export { DAFF_ARTICLE_COMPONENTS } from './article';
 
 export { DaffArticleComponent } from './article/article.component';
+export { DaffArticleCopyButtonComponent } from './article-copy-button/copy-button/copy-button.component';
```
This should not be in the public api.

damienwebdev commented on libs/design/article/src/article.ts:
```diff
@@ -1,4 +1,5 @@
 import { DaffArticleComponent } from './article/article.component';
+import { DaffArticleCopyButtonComponent } from './article-copy-button/copy-button/copy-button.component';
```
This should not be in the public api.

damienwebdev commented on libs/design/article/src/article.module.ts:
```diff
@@ -11,10 +12,12 @@ import { DaffArticleMetaDirective } from './article-meta/article-meta.directive'
   imports: [
     CommonModule,
     DaffArticleComponent,
+    DaffArticleCopyButtonComponent,
```
This module is deprecated and doesn't need this change.

damienwebdev commented on libs/design/article/src/article/article.component.ts:
```diff
@@ -18,5 +25,56 @@ import {
+      document?.defaultView &&
+      typeof document.defaultView.MutationObserver !== 'undefined'
+    ) {
+      observer = new document.defaultView.MutationObserver(mutations => {
```
This is very confusing. I don't think MutationObserver is the right mechanism for this. I would prefer to use contentChildren signals.

joannalauu replied:
Sounds good. I initially used MutationObserver instead of contentChildren because it is hard to track elements injected via `[innerHTML]`, but I'll work on fixing this!!

joannalauu replied:
Hey, I've looked into using contentChildren signals to track content inside `daff-article`, and I don't think this is possible because contentChildren can only track components within the Angular component tree, but the content inside `daff-article` is not part of the Angular component tree since it is rendered through `[innerHTML]`. However, please let me know if I am thinking about this issue incorrectly!!

If the use of MutationObserver isn't ideal, we could revert back to commit [`0ae8bb0`](https://github.com/graycoreio/daffodil/pull/4176/files/0ae8bb0e8330a9909e527ddc752802a9e780a63f), where the copy button is implemented in `docs` package and then imported into daffio.

damienwebdev replied:
That's a very good point that I hadn't considered. I haven't had a chance to dig further into this yet, but that's a very good catch.

damienwebdev commented on libs/design/article/src/article-copy-button/copy-button/copy-button.component.ts:
```diff
@@ -0,0 +1,66 @@
+      await navigator.clipboard.writeText(this.content());
+      this.copied.set(true);
+
+      this.timeoutId = setTimeout(() => {
```
This timeout should be run outside the ngZone.

damienwebdev commented on apps/daffio/src/app/docs/guides/components/guides-content/guides-content.component.ts:
```diff
@@ -34,7 +34,9 @@ export class DaffioDocsGuidesContentComponent implements DaffioDocsDynamicConten
   ) {
     effect((onCleanup) => {
       this.tocRegistry.set(this.doc().tableOfContents);
-      onCleanup(() => this.tocRegistry.set([]));
```
Remove this diff.

damienwebdev commented on apps/daffio/src/app/docs/design/components/component-content/component-content.component.ts:
```diff
@@ -36,6 +36,7 @@ import { DaffioDocsTocHeaderDirective } from '../../../toc/header.directive';
 import { DaffioDocsTocService } from '../../../toc/toc.service';
 import { DaffioDocsDesignApiSortSectionLabels } from '../../pipes/sort-api-section-labels.pipe';
 
+
```
Remove this diff.

damienwebdev reviewed (changes requested):
I'm going to take a stab at this by basing off your branch. There are aspects of this implementation that I find cumbersome and hard to understand.

joannalauu committed:
remove diff

joannalauu committed:
remove copy button from public api

joannalauu committed:
put back doc spec

joannalauu committed:
run timeout outside ngzone

joannalauu committed:
fix linting error

damienwebdev committed:
feat(design): rework article copy-button insert behavior

- adjust copy-button to use daff-icon-button
- add specs for daffio scenario with innerhtml
- rework article to use afterNextRender over MutationObserver
- Adjust article scss to use position:relative on pre and overflow scroll on code
- add some specs for good measure

damienwebdev commented:
@sunray4 I've added a diff to your branch, please take a look:

In short:

- adjust copy-button to use daff-icon-button
- add specs for daffio scenario with innerhtml
- rework article to use afterNextRender over MutationObserver
- Adjust article scss to use position:relative on pre and overflow scroll on code
- add some specs for good measure
- added a timeout clear so that clicking "again" before the timeout doesn't cause strange UI behavior

Notably, I was able to remove the mutation observer using `afterNextRender` in the `write` phase (along with a small check to see if a `pre` already has an associated button`).

I disliked the `wrapper` component as its unnecessary DOM manipulation that can be avoided.

The specs I added specifically cover the scenario in daff.io, so those should eliminate any confusion about the specific scenario.

Some small comments for @xelaint 

 - This PR adjusts the styles from the `pre` element to the `code` element, I'd like your thoughts on this. 
 - I changed the `daff-icon-button` color to `primary`. This probably isn't ideal, however, on small screens the icon button has no background, so its blends into any text underneath it when the codeblock is scrolled.

As per an earlier comment @sunray4 :

> DaffArticleCopyButtonComponent theming is implemented from scratch because importing DaffIconComponent into DaffArticleCopyButtonComponent breaks www.daff.io/docs/design/components/button

I didn't see this anymore. Do you still?

joannalauu commented:
@damienwebdev This looks great, thank you for helping with this PR!!

> > DaffArticleCopyButtonComponent theming is implemented from scratch because importing DaffIconComponent into DaffArticleCopyButtonComponent breaks [www.daff.io/docs/design/components/button](http://www.daff.io/docs/design/components/button)
> 
> I didn't see this anymore. Do you still?

When I serve daffio on my device, [www.daff.io/docs/design/components/button](http://www.daff.io/docs/design/components/button) currently returns 404. I've ran `npm ci` and rebuilt all packages before serving. Running the website in incognito also doesn't work. If the page works on your device, then it's likely a problem on my end, but perhaps double check by running on a third device just to be sure?

xelaint commented on libs/design/article/src/article/article.component.scss:
```diff
@@ -100,18 +100,23 @@
 		code {
-			padding: 0;
+			white-space: pre;
+			-webkit-overflow-scrolling: touch;
```
Remove. It's an invalid property and no longer needed as of safari 13 (https://developer.apple.com/documentation/safari-release-notes/safari-13-release-notes)

xelaint commented on libs/design/article/src/article/article.component.scss:
```diff
@@ -100,18 +100,23 @@
+			overflow-x: auto;
+			display: block;
+			max-width: 100%;
+			min-width: 100px;
```
Don't think these styles are needed.

xelaint commented on libs/design/article/src/article/article.component.scss:
```diff
@@ -100,18 +100,23 @@
-		margin: 1.5rem 0;
-		padding: 1.5rem;
-		overflow-x: auto;
+		overflow: hidden;
```
`overflow: hidden;` is not needed

xelaint commented on libs/design/article/src/article-theme.scss:
```diff
@@ -36,6 +36,14 @@
 				}
 			}
 
+			&__copy-button {
```
for light:
```
background: daff-color($neutral, 10);
border: 1px solid daff-color($neutral, 20);
```

for dark:
```
background: daff-color($neutral, 90);
border: 1px solid daff-color($neutral, 80);
```

xelaint commented on libs/design/article/src/article-copy-button/copy-button/copy-button.component.html:
```diff
@@ -0,0 +1,13 @@
+  daff-icon-button
+  class="daff-article__copy-button"
+  type="button"
+  color="primary"
```
Remove `color="primary"`

xelaint commented on libs/design/article/src/article-copy-button/copy-button/copy-button.component.scss:
```diff
@@ -0,0 +1,9 @@
+@use '../../../../scss/interactions';
```
Remove. This is not used.

xelaint commented on libs/design/article/src/article-copy-button/copy-button/copy-button.component.scss:
```diff
@@ -0,0 +1,9 @@
+	display: block;
+	position: absolute;
+	top: 0.2rem;
+	right: 0.2rem;
```
```
top: 0.25rem;
right : 0.25rem;
```

xelaint commented on libs/design/article/src/article/article.component.scss:
```diff
@@ -100,18 +100,23 @@
 
 		code {
-			padding: 0;
+			white-space: pre;
```
Can you reorganize this code with display styles first, with the rest listed alphabetically?

joannalauu committed:
style fixes

damienwebdev committed:
Add @daffodil/design to devDependencies

damienwebdev reviewed (approved):
lgtm
